src/CC.hs

```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE RecordWildCards #-}
+-- |
+--
+-- Utility for formatting @'Idea'@ data in accordance with the Code Climate
+-- spec: <https://github.com/codeclimate/spec>
+--
+module CC
+    ( printIssue
+    , fromIdea
+    ) where
+
+import Data.Aeson (ToJSON(..), (.=), encode, object)
+import Data.Char (toUpper)
+import Data.Monoid ((<>))
+import Data.Text (Text)
+import Language.Haskell.Exts.SrcLoc (SrcSpan(..))
+
+import qualified Data.Text as T
+import qualified Data.ByteString.Lazy.Char8 as C8
+
+import Idea (Idea(..), Severity(..))
+
+data Issue = Issue
+    { issueType :: Text
+    , issueCheckName :: Text
+    , issueDescription :: Text
+    , issueContent :: Text -- TODO: Markdown type?
+    , issueCategories :: [Text]
+    , issueLocation :: Location
+    , issueRemediationPoints :: Int
+    }
+
+data Location = Location FilePath Position Position
+data Position = Position Int Int
+
+instance ToJSON Issue where
+    toJSON Issue{..} = object
+        [ "type" .= issueType
+        , "check_name" .= issueCheckName
+        , "description" .= issueDescription
+        , "content" .= object
+            [ "body" .= issueContent
+            ]
+        , "categories" .= issueCategories
+        , "location" .= issueLocation
+        , "remediation_points" .= issueRemediationPoints
+        ]
+
+instance ToJSON Location where
+    toJSON (Location path begin end) = object
+        [ "path" .= path
+        , "positions" .= object
+            [ "begin" .= begin
+            , "end" .= end
+            ]
+        ]
+
+instance ToJSON Position where
+    toJSON (Position line column) = object
+        [ "line" .= line
+        , "column" .= column
+        ]
+
+-- | Print an @'Issue'@ with trailing null-terminator and newline
+--
+-- The trailing newline will be ignored, but makes the output more readable
+--
+printIssue :: Issue -> IO ()
+printIssue = C8.putStrLn . (<> "\0") . encode
+
+-- | Convert an hlint @'Idea'@ to a datatype more easily serialized for CC
+fromIdea :: Idea -> Issue
+fromIdea Idea{..} = Issue
+    { issueType = "issue"
+    , issueCheckName = "HLint/" <> T.pack (camelize ideaHint)
+    , issueDescription = T.pack ideaHint
+    , issueContent = content ideaFrom ideaTo
+    , issueCategories = categories ideaHint
+    , issueLocation = fromSrcSpan ideaSpan
+    , issueRemediationPoints = points ideaSeverity
+    }
+
+  where
+    content from Nothing = T.unlines
+        [ "Found"
+        , ""
+        , "```"
+        , T.pack from
+        , "```"
+        , ""
+        , "remove it."
+        ]
+
+    content from (Just to) = T.unlines
+        [ "Found"
+        , ""
+        , "```"
+        , T.pack from
+        , "```"
+        , ""
+        , "Why not"
+        , ""
+        , "```"
+        , T.pack to
+        , "```"
+        ]
+
+    categories _ = ["Style"]
+
+    points Ignore = 0
+    points Suggestion = basePoints
+    points Warning = 5 * basePoints
+    points Error = 10 * basePoints
```

Comment:
It seems like these points are markers of difficulty rather than importance - so are these numbers sensible?

Reply:
I would love you input on that. The comment down with `basePoints` describes how it's meant to work, using that could you recommend rough numbers for the various severities?